test/GPv2Transfer.test.ts

```diff
@@ -0,0 +1,160 @@
+import IERC20 from "@openzeppelin/contracts/build/contracts/IERC20.json";
+import { expect } from "chai";
+import { MockContract } from "ethereum-waffle";
+import { Contract } from "ethers";
+import { artifacts, ethers, waffle } from "hardhat";
+
+import { BUY_ETH_ADDRESS } from "../src/ts";
+
+describe("GPv2Transfer", () => {
+  const [deployer, recipient, ...traders] = waffle.provider.getWallets();
+
+  let transfer: Contract;
+  let vault: MockContract;
+  let token: MockContract;
+
+  beforeEach(async () => {
+    const GPv2Transfer = await ethers.getContractFactory(
+      "GPv2TransferTestInterface",
+    );
+    transfer = await GPv2Transfer.deploy();
+
+    const IVault = await artifacts.readArtifact("IVault");
+    vault = await waffle.deployMockContract(deployer, IVault.abi);
+    token = await waffle.deployMockContract(deployer, IERC20.abi);
+  });
+
+  describe("transferToRecipient", () => {
+    const amount = ethers.utils.parseEther("13.37");
+
+    it("should transfer external amount to recipient", async () => {
+      await token.mock.transferFrom
+        .withArgs(traders[0].address, recipient.address, amount)
+        .returns(true);
+      await expect(
+        transfer.transferToRecipientTest(vault.address, recipient.address, [
+          {
+            account: traders[0].address,
+            token: token.address,
+            amount: amount,
+            useInternalBalance: false,
+          },
+        ]),
+      ).to.not.be.reverted;
+    });
+
+    it("should transfer internal amount to recipient", async () => {
+      await vault.mock.withdrawFromInternalBalance
+        .withArgs([
+          {
+            token: token.address,
+            amount,
+            sender: traders[0].address,
+            recipient: recipient.address,
+          },
+        ])
+        .returns();
+      await expect(
+        transfer.transferToRecipientTest(vault.address, recipient.address, [
+          {
+            account: traders[0].address,
+            token: token.address,
+            amount,
+            useInternalBalance: true,
+          },
+        ]),
+      ).to.not.be.reverted;
+    });
+
+    it("should transfer many external and internal amounts to recipient", async () => {
+      const transfers = traders.map((trader, i) => ({
+        account: trader.address,
+        token: token.address,
+        amount,
+        useInternalBalance: (i & 1) == 1,
+      }));
+
+      for (const { account } of transfers.filter(
+        (transfer) => !transfer.useInternalBalance,
+      )) {
```

Comment:
Just to be sure, I'd like to see some assertions that both filters let some elements pass.